### ISSUE TITLE
How to exclude a directory specified with parent dirs?

### ISSUE THREAD

**Issue by kaushalmodi** [universal-ctags/ctags#759]
Hi,

I believe that the below in ~/.ctags used to work:

--exclude=dir1/dir2/common

But it is not working any more as of the latest build of today (My last ctags build was roughly 3 months back. So unfortunately that will not help in debugging this).

If I change that to 

--exclude=common

it works. But the side-effect is that all common/ directories will be excluded, which I don't want.

Example:

I have these directory structures:

dir1/dir2/common
dir1/dir3/common

How can I exclude only dir1/dir2/common?

Thanks.

**Comment by masatake** [comment]
How about `--exclude=dir1/dir2/common/*` ?

**Comment by masatake** [comment]
NOTE: isExcludedFile@(ctags/main/options.c)

**Comment by masatake** [comment]
dir0/dir1/dir2/common and ./dir1/dir2/common cannot be excluded with `--exclude=dir1/dir2/common/*`
`--exclude=*/dir1/dir2/common/*` may be needed here.

**Comment by kaushalmodi** [comment]
> How about --exclude=dir1/dir2/common/* ?

That works or doesn't work based on the invocation of ctags. See explanation below.

> How did you invoke ctags?

```
cd /proj/PRJ/somedir/
ctags -Re -f /proj/PRJ/somedir/TAGS /proj/PRJ/somedir/.
```

---


If I do the above style of ctags invocation, `--exclude=dir1/dir2/common/*` does not work. But if I invoke using `cd /proj/PRJ/somedir/; ctags -Re -f TAGS .`, the same exclude statement works as expected.

**Comment by kaushalmodi** [comment]
Also, I did not understand this:

> NOTE: isExcludedFile@(ctags/main/options.c)

**Comment by kaushalmodi** [comment]
Another strange behavior I just noticed:

With `--exclude=*/dir1/dir2/common/*`,

Below works

```
cd /proj/PRJ/somedir/
ctags -Re -f /proj/PRJ/somedir/TAGS /proj/PRJ/somedir/.
```

But below does not!

```
cd /proj/PRJ/somedir/
ctags -Re -f /proj/PRJ/somedir/TAGS .
```

Why is the exclusion sensitive to whether the TAGS file path is specified relatively or using the absolute path?

**Comment by masatake** [comment]
> Why is the exclusion sensitive to whether the TAGS file path is specified relatively or using the absolute path?

I don't understand the situation.
Where is dir1 in /proj/PRJ/somedir ?

**Comment by kaushalmodi** [comment]
I am doing some more testing .. here is what I got so far:

```
# --exclude=*/dir1/dir2/common/*
# Good - ctags -Re -f /proj/PRJ/somedir/TAGS /proj/PRJ/somedir/.
# Good - ctags -Re -f TAGS                   /proj/PRJ/somedir/.
# Bad  - ctags -Re -f /proj/PRJ/somedir/TAGS .
# Bad  - ctags -Re -f TAGS .
# Conclusion: When using * in the beginning of an exclusion path, use absolute
#             path for search start path when invoking ctags.
```

> Where is dir1 in /proj/PRJ/somedir ?
Here is the `dir1` path: `/proj/PRJ/somedir/dir1`

**Comment by masatake** [comment]
Oh, I see. Thank you. For the Bad cases `--exclude=dir1/dir2/common/*` will work.
... How about `--exclude=*dir1/dir2/common/*` ? This will work well for the both Good and Bad patterns.
Sure, xdir1/dir2/common is also excluded unexpectedly.

**Comment by kaushalmodi** [comment]
> How about --exclude=*dir1/dir2/common/* ? This will work well for the both Good and Bad patterns.

Yes, that's the conclusion I came to finally :) Should I close this issue? Or do you think there's a solution that make this more robust whether the user is using absolute or relative paths?

For example, if the user specifies a path starting with `/`, it should be detected as an absolute path and the exclude pattern should be appended to it directly.

If the user's start path is NOT start with `/` but is some relative path XYZ, the exclude pattern should be appended to `$cwd/XYZ/`.

**Comment by vkottler** [comment]
Scenario:
1. Want to tag files under `src` (`--languages=C,C++`)
2. Want to exclude files under `src/third-party`

I have come back to this tool so many times, and every single time I do it's the `--exclude` that never works, no matter how many trial-and-error attempts are made to get it to ignore even a single directory.

```
$ ctags --version
Universal Ctags 6.0.0(dbeed621), Copyright (C) 2015-2022 Universal Ctags Team
Universal Ctags is derived from Exuberant Ctags.
Exuberant Ctags 5.8, Copyright (C) 1996-2009 Darren Hiebert
  Compiled: May 13 2023, 18:33:30
  URL: https://ctags.io/
  Output version: 0.0
  Optional compiled features: +wildcards, +regex, +iconv, +option-directory, +packcc, +optscript
```

Every single combination/permutations of arguments do nothing:

```
ctags --languages=C,C++ -R src \
    --exclude=third-party \
    --exclude=src/third-party \
    --exclude=*src/third-party \
    --exclude=src/third-party* \
    --exclude=*src/third-party* \
    --exclude=**src/third-party \
    --exclude=src/third-party** \
    --exclude=*/src/third-party
...
```

Don't feel like typing out the remaining 20-30 option permutations.

Why doesn't `--exclude` just match a sub-string in the path, ignore it and move on? Seriously confusing and frustrating.

I have a few very fragile `ctags` setups that work for me but there's always the fear/dread that the next time I need to use it, this will happen. Just not understanding.
